users/models.py

```diff
@@ -1,3 +1,287 @@
+    is_staff = models.BooleanField(default=False)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    is_soft_deleted = models.BooleanField(default=False)
+    deleted_at = models.DateTimeField(blank=True, null=True)
+    last_login = models.DateTimeField(null=True, blank=True)
+    original_data = models.JSONField(null=True, blank=True)
+
+    objects = CustomUserManager()
+
+    USERNAME_FIELD = 'email'
+    REQUIRED_FIELDS = ['first_name', 'last_name', 'user_phone']
+
+    def __str__(self):
+        """Return a string representation of the user."""
+        return self.email
+
+    def get_full_name(self):
+        """Return the user's full name."""
+        first_name = self.first_name.strip() or 'Unknown first name'
+        last_name = self.last_name.strip() or 'Unknown last name'
+        return f"{first_name} {last_name}".strip()
+
+    def get_roles_display(self):
+        """Return a string representation of the user's roles."""
+        return ', '.join(role.name for role in self.roles.all())
+
+    def add_role(self, role_name):
+        """
+        Add a role to the user.
+
+        Args:
+            role_name (str): Name of the role to add.
+
+        Raises:
+            ValidationError: If the role is invalid or already assigned.
+        """
+        if role_name not in self.ALLOWED_ROLES:
+            raise ValidationError(f"Cannot add role: {role_name}")
+        if not self.is_active:
+            raise ValidationError("This account is not active.")
+        if self.is_soft_deleted:
+            raise ValidationError("This account is deleted.")
+        try:
+            role = Role.objects.get(name=role_name)
+        except Role.DoesNotExist:
+            raise ValidationError(f"Role '{role_name}' does not exist.")
+        if self.roles.filter(name=role_name).exists():
+            raise ValidationError(f"You already have the role: {role_name}")
+        self.roles.add(role)
+        self.save()
+        logger.info(f"Role '{role_name}' added to user {self.email}")
+
+    def remove_role(self, role_name):
+        """
+        Remove a role from the user.
+
+        Args:
+            role_name (str): Name of the role to remove.
+
+        Raises:
+            ValidationError: If the role is invalid or not assigned.
+        """
+        if role_name not in self.ALLOWED_ROLES:
+            raise ValidationError(f"Cannot remove role: {role_name}")
+        if not self.is_active:
+            raise ValidationError("This account is not active.")
+        if self.is_soft_deleted:
+            raise ValidationError("This account is deleted.")
+        try:
+            role = self.roles.get(name=role_name)
+        except Role.DoesNotExist:
+            raise ValidationError(f"User does not have the role: {role_name}")
+        self.roles.remove(role)
+        self.save()
+        logger.info(f"Role '{role_name}' removed from user {self.email}")
```

Comment:
The `add_role` and `remove_role` methods work well, but you can simplify them by directly using the `self.roles` manager to get or add roles.
Combine the existence checks with the operation to avoid extra queries.

Reply:
done